### ISSUE TITLE
Register `VersionCatalogsExtension` before `beforeProject` lifecycle hooks

### ISSUE THREAD

**Issue by TheMrMilchmann** [gradle/gradle#31289]
### Expected Behavior

As per documentation of `Gradle::beforeProject`, the block is called immediately before the project is evaluated. I would expect version catalogs (via `VersionCatalogsExtension`) to be available in this block. Since version catalogs are built from settings, I don't see an obvious reason for why they would not be ready there.

Specifically, I would expect this (and similar) logic to work (in `settings.gradle.kts`):

```kotlin
gradle.lifecycle.beforeProject {
    extensions.getByType(VersionCatalogsExtension::class)
}
```

### Current Behavior (optional)

In a project plugin registered via settings plugin in `beforeProject`, I cannot access version catalogs because the extension is not registered at this stage. The snippet above fails with `Extension of type 'VersionCatalogsExtension' does not exist.`.

### Context

I have multiple repositories that have the same subproject layout. To deduplicate build logic, I planned to create a settings plugin that does the following:
1. Register an external version catalog
2. Including the usual subprojects
3. Applying project plugins to each subproject

The project plugins would have used dependencies from the catalog - and this is where the idea fails due to the current behavior.

**Comment by ov7a** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by mlopatkin** [comment]
This feature request is in the backlog of the relevant team and is prioritized by them.

***

We should check if the legacy `beforeProject` or `allprojects` callbacks can access this extension. It makes sense to have the extension available in both.

**Comment by TheMrMilchmann** [comment]
> We should check if the legacy `beforeProject` or `allprojects` callbacks can access this extension. It makes sense to have the extension available in both.

If you're referring to the `Gradle::beforeProject` and `Gradle::allProjeccts` methods, I checked and they suffer from the same issue.

**Comment by stack-smash** [comment]
Yes, `allprojects` also suffers from the extension not being available. I tried using the `allprojects` block in an init script and got the same error. [This](https://discuss.gradle.org/t/version-catalogs-and-init-scripts/50173) is what I was trying out.
